### PR TITLE
Lua api for autopilot functions

### PR DESCRIPTION
**Pull request by maxim-productengine**
Also made some tweaks and fixed detection of appropriate nearby object (we don't need attachments in all cases, but need extra check if we're looking for sit target).
When autopilot is terminated, success/fail status is posted to the event pump, which can be used by the script.
Changed LLChatListener to more general LLListener, which can be used for setting any pump listener, instead of just nearby chat.

### CODE REVIEW

**Review by nat-goodspeed — CHANGES_REQUESTED** [review]
I like most of this. The change I want to see is in the new generalized `LLListener`. Thanks!

File: indra/newview/llagentlistener.cpp [hunk 1]
```diff
@@ -87,14 +87,14 @@ LLAgentListener::LLAgentListener(LLAgent &agent)
     add("startAutoPilot",
         "Start the autopilot system using the following parameters:\n"
         "[\"target_global\"]: array of target global {x, y, z} position\n"
-        "[\"stop_distance\"]: target maxiumum distance from target [default: autopilot guess]\n"
+        "[\"stop_distance\"]: target maximum distance from target [default: autopilot guess]\n"
```
[the remark this answers is not in the document]
Comment (maxim-productengine) [reply indra/newview/llagentlistener.cpp:90]: Changed this description to 'maximum stop distance from target'.

File: indra/newview/llagentlistener.cpp [hunk 2]
```diff
@@ -463,20 +483,27 @@ void LLAgentListener::startFollowPilot(LLSD const & event_data)
...
     }
 
-    if (target_id.notNull())
+    if (!gObjectList.findObject(target_id))
     {
-        mAgent.setFlying(allow_flying);
-        mFollowTarget = target_id;  // Save follow target so we can report distance later
-
-        mAgent.startFollowPilot(target_id, allow_flying, stop_distance);
+        std::string target_info = event_data.has("leader_id") ? event_data["target_id"] : event_data["avatar_name"];
```
[the remark this answers is not in the document]
Comment (maxim-productengine) [reply indra/newview/llagentlistener.cpp:499]: Thanks, it should be `event_data["leader_id"]`

File: indra/newview/scripts/lua/require/LLAgent.lua [hunk 3]
```diff
@@ -80,4 +80,64 @@ function LLAgent.teleport(...)
...
+--  Autopilot
+-- ***************************************************************************
+LLAgent.autoPilotPump = "LLAutopilot"
+
+-- Start the autopilot to move to "target_global" location using specified parameters
+-- LLAgent.startAutoPilot{ target_global       array of target global {x, y, z} position
+--                      [, allow_flying]       allow flying during autopilot [default: true]
+--                      [, stop_distance]      target maximum distance from target [default: autopilot guess]
+--                      [, behavior_name]      name of the autopilot behavior [default: (script name)]
+--                      [, target_rotation]    array of [x, y, z, w] quaternion values [default: no target]
```
[the remark this answers is not in the document]
Comment (maxim-productengine) [reply indra/newview/scripts/lua/require/LLAgent.lua:108]: Yes, exactly, `target_rotation ` specifies desired avatar facing direction, for example related to our map:
target_rotation={0,0,-0.71,0.71} -- facing South
target_rotation={0,0,0.71,0.71} -- facing North
target_rotation={0,0,0,1} -- facing East
target_rotation={0,1,0,0}-- facing West

Basically, we're rotating x_axis (1, 0, 0) vector to the desired rotation using this quaternion.

**Pull request closed without merging**